Add features and organize code to be more customizable

zhshize opened the pull request:
I think that **Customizable** is the most important property when I am applying this library to many old projects.
So I do many modifications, it will be fantastic!

I merged new updates during development, don't worry about the inconsistency between two repos.

The doc is added in README.md, it is helpful.

I think that the API changes are so significant that it **could be a major version release**.

# What's inside

Fixes:
 - Typos

Enhancements:
 - Type declarations
 - Documentation & PHPDocs
 - Add PHP version requirement
 - Move the interfaces to Contacts namespace (shows that these interfaces are good to implement and replace)
 - Move exception classes to Exceptions namespace
 - Add events, all OIDC and authentication related events will be dispatched
 - Add `Contracts\OIDCAuthenticatable` to mark user object with access token features
 - Split `OIDCAuthServiceProvider` into many tiny functions for customization (extends and small overwritings)
 - Split `CallbackController` into many tiny functions for customization (extends and small overwritings)

Features:
 - Refreshing access token (a function for refresh manually and an option for refresh automatically)
 - Add Facade `OIDC`

Modifications:
 - New OIDCSessionGaurd (doesn't extend Laravel SessionGuard for being concise)
 - All codes about authorization code flow are moved into the class `AuthorizationCodeFlow`
 - All codes about storing token are moved into the class `TokenStore`

# Significant Updates

 - `config/oidc-auth.php` **renamed** to `config/oidc_auth.php`  
 Because of Laravel naming convention
 - `UserFactoryInterface` moved to `Contracts\OIDCAuthenticatableFactory`
 - Middleware `Authenticate` moved to `Middleware\Authenticate`
 - `SessionGuard` **replaced** by `OIDCSessionGaurd `
 - `SessionUserProvider` renamed to `OIDCSessionUserProvider`
 - `InvalidStateException`, `AuthenticationException`, `AuthenticationErrorException` moved to Exceptions namespace
 
 Give me some comment, I will be glad to improve the code.

xdavidwu commented:
> I think that the API changes are so significant that it could be a major version release.

Per SemVer, we won't bump major unless we feel easy to gurantee API stability.

> config/oidc-auth.php renamed to config/oidc_auth.php
> Because of Laravel naming convention

Is that a thing in `config/` naming? A quick `find` in my work environment over `vendor/` of a few laravel projects shows a case with `-` (ide-helper.php in barryvdh/laravel-ide-helper, I believe this is popular), and nothing with a `_`.

This is a huge patchset and require time to review. You should probably split the changes into individual PRs next time.

xdavidwu commented on composer.json:
```diff
@@ -15,6 +15,7 @@
         }
     ],
     "require": {
+        "php": "^7.4|^8.0",
```
Do we need to be at least 7.4? The most modern addition in PHP we use, iirc, is null coalescing operator, which is available since 7.0.

xdavidwu replied:
Oh I see you use typed properties (7.4+) in follow-ups

zhshize replied:
oauth2-openid-connect-client requires 7.4+

https://github.com/steverhoades/oauth2-openid-connect-client#requirements

xdavidwu replied:
It is not needed to bring up used library dependency. That is package manager's job.

The real reason for doing this should be:
> you use typed properties (7.4+) in follow-ups

xdavidwu commented on src/OIDCSession.php:
```diff
@@ -0,0 +1,47 @@
+{
+    private string $sessionKey = 'oidc-auth.access_token';
+
+    public function user(): ?OIDCUser
```
I don't think this one will be useful

zhshize replied:
Removed

xdavidwu commented on src/AuthorizationCodeFlow.php:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    /**
+     * After the OpenID provider returned an Authentication Response, call this function with Request
```
OpenID Connect != OpenID

zhshize replied:
'OpenID provider' is a term in spec.

https://openid.net/specs/openid-connect-core-1_0.html#Terminology

zhshize replied:
Change to 'OpenID Connect provider' in newer commit

xdavidwu commented on src/OIDCSessionGuard.php:
```diff
@@ -0,0 +1,101 @@
+                        return false;
+                    }
+                } else {
+                    $this->OIDCSession->clearAccessToken();
```
side-note: This is a behavoir change that enforce expiration validation futher in already authenticated session lifecycle.

zhshize replied:
Downstream user won't need to validate the expiration, I think that it is a useful change.

xdavidwu replied:
I'm not against it. I am just noting it for further release planning.

Please keep this kind of side-noting conversation open.

xdavidwu commented on config/oidc-auth.php:
```diff
@@ -41,7 +41,7 @@
     |
     | Factory to get an Illuminate\Contracts\Auth\Authenticatable to use, see
-    | LaravelOIDCAuth\UserFactoryInterface.
+    | LaravelOIDCAuth\Contracts\UserFactoryInterface.
```
Not fully renamed here

zhshize replied:
fixed in newer commit

xdavidwu commented on src/OIDCSessionGuard.php:
```diff
@@ -37,6 +37,8 @@ public function user(): ?Authenticatable
             }
             return $this->user;
+        } else {
+            $this->OIDCSession->clearAccessToken();
```
When `tryRefreshToken()` returns false, there is already no token, either cleared by it or not there in the first place.

xdavidwu commented on src/OIDCAuthServiceProvider.php:
```diff
@@ -2,41 +2,88 @@
+    /**
+     * Register OIDC Auth services.
+     *
+     * Please override other functions in this class instead of boot() to customize.
```
I don't think it is useful to customize this class.  I don't think any reason doing this, as it just registers stuff and not actually enforce using them.

zhshize replied:
Refactored in newer commit

xdavidwu commented on src/CallbackController.php:
```diff
@@ -2,29 +2,100 @@
+    /**
+     * @var bool Indicated whether refresh the access token when token expired
+     */
+    protected bool $autoRefresh = false;
```
This should come from config.

Also I'm not interested in maintaining "customizability" in this class. I don't see valuable reasons doing so, and I am not a fan of making every tiny step a method. The logic is already clear.

zhshize replied:
Refactored in newer commit

xdavidwu commented on src/Events/OIDCAuthenticated.php:
```diff
@@ -0,0 +1,31 @@
+use Illuminate\Queue\SerializesModels;
+
+/**
+ * Represent the event when OpenID Provider authenticated and sent the authorization code
```
OpenID Connect != OpenID

zhshize replied:
fixed in newer commit

xdavidwu commented on src/Events/OIDCAuthenticated.php:
```diff
@@ -0,0 +1,31 @@
+    use Dispatchable, InteractsWithSockets, SerializesModels;
+
+    /**
+     * @var Request|null The request returned by OpenID Provider after authorization redirected
```
ditto

zhshize replied:
fixed in newer commit

xdavidwu commented on src/Events/OIDCAuthenticationFailed.php:
```diff
@@ -0,0 +1,35 @@
+    use Dispatchable, InteractsWithSockets, SerializesModels;
+
+    /**
+     * @var Request|null The request returned by OpenID Provider after authorization redirected
```
ditto

zhshize replied:
fixed in newer commit

xdavidwu commented on src/Events/OIDCAuthenticationStarted.php:
```diff
@@ -0,0 +1,33 @@
+    public string $authorizationUrl;
+
+    /**
+     * @var string Opaque value used to maintain state between the request and the callback.
```
Sounds vague. I think if you want to explain, you should just link to related section of OIDC/OAuth2 spec/RFC.

zhshize replied:
Modify the comment

zhshize replied:
fixed in newer commit

xdavidwu commented on src/AccessTokenRefreshable.php:
```diff
@@ -0,0 +1,33 @@
+ * @see HasAccessToken
+ * @see OIDCSessionGuard
+ */
+trait AccessTokenRefreshable
```
Naming: RefreshesAccessToken?

zhshize replied:
'RefreshesAccessToken' is better, I'll rename it.

xdavidwu reviewed (commented):
First round of review until `d28bc0642d712fad2ebec9ae1d0808a2cc15ac35`

I see you changing your design over and over again within this PR. There are somethings that I agree with, but also some I have my doubt. I don't see convincing reasons in some of your focus on "customizabilty", and I'm afraid some of those changes reduce simplicity and make this code base less maintainable.

zhshize commented on config/oidc_auth.php:
```diff
@@ -30,25 +30,48 @@
-    | Factory to get a Illuminate\Contracts\Auth\Authenticatable to use, see
-    | LaravelOIDCAuth\UserFactoryInterface.
-    | For example, you can use a Eloquent model as Authenticatable to store
+    | Factory to get an Illuminate\Contracts\Auth\Authenticatable to use, see
```
OIDCAuthenticatable

zhshize replied:
fixed in newer commit

zhshize commented:
To Do:
- List breaking changes
- Fold CallbackController, TokenStore, OIDCAuthServiceProvider
- Remove OIDC Facade
- Other issues in above conversations
- Rearrange commit history
- Rebase from main

zhshize committed:
Add php version requirement 7.4 or 8.0

zhshize committed:
Add user & user factory contracts

- Add OIDCAuthenticatable
- Replace UserFactoryInterface with OIDCAuthenticatableFactory

zhshize committed:
Add events

zhshize committed:
Add AuthorizationCodeFlow

zhshize committed:
Refactor CallbackController with AuthorizationCodeFlow

zhshize committed:
Move exceptions

zhshize committed:
Misc changes for config and route

- Fix typo
- Use https in example
- Rename config option

zhshize committed:
Refactor Authenticate middleware

zhshize committed:
Add OIDCSessionGuard to replace SessionGuard

zhshize committed:
Refactor OIDCUser

zhshize committed:
Refactor OIDCAuthServiceProvider.php

zhshize committed:
Update README.md

zhshize committed:
Update composer.lock

zhshize commented:
### Breaking changes
#### Renamed & moved files
- `config/oidc-auth.php` renamed to `config/oidc_auth.php`
- Exceptions was moved from top level namespace `LaravelOIDCAuth` to `LaravelOIDCAuth\Exceptions`:
  - `AuthenticationErrorException`
  - `AuthenticationException`
  - `InvalidStateException`
- Middleware `LaravelOIDCAuth\Authenticate` moved to `LaravelOIDCAuth\Middleware\Authenticate`
#### Removed files
- Class `OIDCProviderService` was removed
- Class `SessionUserProvider` was removed
#### UserFactory
The UserFactory must implement `LaravelOIDCAuth\Contracts\OIDCAuthenticatableFactory`. `UserFactoryInterface` was removed.

xdavidwu commented:
about addition of composer.lock: I think we don't actually need to use lock file in library development. It will not be respected when used as a library anyway. I am against tracking a lock file as it may hinder potential chances on catching problems on development.

xdavidwu commented:
> config/oidc-auth.php renamed to config/oidc_auth.php

I'm still against this until this convention is documented in official documents, or there is widespread adoption.

>  LaravelOIDCAuth\OIDCAuthenticatableFactory

afaict it's `LaravelOIDCAuth\Contracts\OIDCAuthenticatableFactory`

xdavidwu commented on README.md:
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
 Set guard driver to `oidc-auth-session` to remove tokens from session storage on logout.
 
-Set provider driver to `oidc-auth-session` to make `Auth::user()` return a `LaravelOIDCAuth\OIDCUser` authenticable from saved token.
+Set provider driver to `none`, user provider is not suitable.
```
> is not suitable

is not needed

Edit: `is not used` should be better

> Set provider driver to `none`

Removing that key may be better

zhshize replied:
I agree.

xdavidwu commented on README.md:
````diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
+```php
+// Get login user, returns \LaravelOIDCAuth\Contracts\OIDCAuthenticatable|null
+Auth::user();
+// Get the saved access token, returns \OpenIDConnectClient\AccessToken|null
````
Is null possible on `accessToken()` (and also those behind that) with default implementation?

xdavidwu commented on README.md:
````diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
+You can get id token claims like this:
+
+```php
+Route::get('/user', function () {
````
For consistency with previous examples, we should not include route definitions.

zhshize replied:
Okay.

xdavidwu commented on README.md:
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
+
+### Refresh the access token
+
+If you are using `oidc-auth-session` guard with `OIDCUser`, you can easily refresh the access token:
```
> easily refresh the access token:

refresh access token like following:

zhshize replied:
Okay.

xdavidwu commented on README.md:
````diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
+If you are using `oidc-auth-session` guard with `OIDCUser`, you can easily refresh the access token:
+
+```php
+use LaravelOIDCAuth\Facades\OIDC;
````
facade not actually used

zhshize replied:
Okay.

xdavidwu commented on README.md:
````diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
+Auth::user()->refreshAccessToken();
+```
+
+Typically, you don't need to refresh the access token manually, the following doc shows an automatic way to do this.
````
> Typically, you don't need to refresh the access token manually, the following doc shows an automatic way to do this.

You can also turn the `Auto refresh`  feature on instead of refreshing manually.

We should also mention that whether token is actually refreshed by that function if it is not expired. If it is always refreshed, we should document how to check expiration against existing token.

Or even better, `Auto refresh` section should come before this one.

zhshize replied:
I rearranged the doc.

xdavidwu commented on README.md:
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
+
+#### Auto refresh
+
+Sometimes, you want the `oidc-auth-session` to refresh the access token automatically every request, please edit
```
If you need to automatically validate expiration and refresh tokens if possible, enable `auto_refresh` in config.

(No code example required as it is trivial.)

zhshize replied:
Okay.

xdavidwu commented on README.md:
````diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
+    // ...
+```
+
+Next time the user logged in, the `CallbackController` will set `rememberMe` option from
````
Is this rememberMe mechanism related to refreshing token? Why is it tied?

zhshize replied:
Yes, it is.

If the user log in with 'remember me', guard will automatically refresh the access token to keep the user logged in when access token is expired.  Actually, the 'auto_refresh' option is treated as the default value of the rememberMe argument when controller log the user in(See: [loginUser](https://github.com/zhshize/laravel-oidc-auth/blob/53bfe0bfbfc47104e9a7ef6e1c6618b79fb2991f/src/CallbackController.php#L91)).  This mechanism is a simulation of remember me feature when we integrate OIDC to Laravel authentication.

Do you think the mechanism should be described in the doc or this mechanism could be confusing?

xdavidwu replied:
Well I see you just removed that in newer commit.

If it is implementation detail, it should not be doc'ed at README. We do not want to gurantee this level of compatibility.

IIRC, "remember me" in Laravel document is some sort of long-lived cookie stuff, which is unrelated to this token refreshing behavior. Better use a more straight-forward naming in code instead of mimicking something that we don't actually implement.

zhshize replied:
I renamed the parameter remember to auto_refresh in the newest commit.

xdavidwu commented on README.md:
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
+Next time the user logged in, the `CallbackController` will set `rememberMe` option from
+`config('oidc_auth.auto_refresh')` to the guard's `login` function.
+
+### User Model & User Factory
```
For consistency for cases, `User model & user factory`?

zhshize replied:
Okay.

xdavidwu commented on README.md:
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
+
+### User Model & User Factory
+
+If you need an authenticated user to be an eloquent model with OIDC features, your user model must implement
```
If a custom user implementation is desired (for example, as an Eloquent model), it should implement ...

zhshize replied:
Okay.

xdavidwu commented on README.md:
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
+
+class User extends Authenticatable implements OIDCAuthenticatable
+{
+    use Notifiable, HasAccessToken, RefreshesAccessToken;
```
Why is Notifiable required?

Also, I don't think code example is required here. Mentioning traits that may help implementation is enough.

zhshize replied:
Okay.

xdavidwu commented on README.md:
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
+{
+    public function authenticatable(AccessToken $token): OIDCAuthenticatable
+    {
+        $user = User::firstOrCreate(/* ... */);
```
Again, no code example should be required. This also assumes an Eloquent model, which may not be the only usecase.

zhshize replied:
Okay.

xdavidwu commented on README.md:
```diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
+
+#### Handle OIDC callback
+
+When the OIDC Provider authenticated the user, the browser will be redirected to the callback route on your own web
```
When user authenticates via OIDC,

Edit: When user is authenticated by OIDC Provider, the browser will be redirected to callback route...

xdavidwu replied:
(I don't really think you should explain callback here. Consumer should already know it if they need to handle that manually)

zhshize replied:
Okay.

xdavidwu commented on README.md:
````diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
+route in `config/oidc_auth.php`:
+
+```php
+    'register_default_callback_route' => false,  // default value is true, set to false
````
Pointing out defaults here is redundant (you are already pointing out this need to be set manually.)

zhshize replied:
Okay.

xdavidwu commented on README.md:
````diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
+Here is an example of your callback controller:
+
+```php
+namespace App\Http\Controllers;
````
Trivial, no code example needed.

zhshize replied:
Okay.

xdavidwu commented on README.md:
````diff
@@ -8,14 +8,157 @@ Based on steverhoades/oauth2-openid-connect-client.
+}
+```
+
+Add a route in `route/web.php`:
````
Also trivial and unnecessarily assumes consumer's route organization.

zhshize replied:
Okay.

xdavidwu reviewed (commented):
A few comments for documentation only

zhshize commented:
> > config/oidc-auth.php renamed to config/oidc_auth.php
> 
> I'm still against this until this convention is documented in official documents, or there is widespread adoption.
> 

There is an example: https://github.com/laravel/cashier-mollie/tree/develop/config
and a non-official guide: https://github.com/alexeymezenin/laravel-best-practices#follow-laravel-naming-conventions

I like `config('oidc_auth.auto_refresh')` better than `config('oidc-auth.auto_refresh')`.

> > LaravelOIDCAuth\OIDCAuthenticatableFactory
> 
> afaict it's `LaravelOIDCAuth\Contracts\OIDCAuthenticatableFactory`

`LaravelOIDCAuth\OIDCAuthenticatableFactory` -> `LaravelOIDCAuth\Contracts\OIDCAuthenticatableFactory` updated.

zhshize commented:
> about addition of composer.lock: I think we don't actually need to use lock file in library development. It will not be respected when used as a library anyway. I am against tracking a lock file as it may hinder potential chances on catching problems on development.

I'll remove that commit.

xdavidwu commented:
> There is an example: https://github.com/laravel/cashier-mollie/tree/develop/config

While that is not a famous package, it seems like that is the only example with multiple words under laravel namespace (although it is moving out). If we are going for underscores, we should also rename the route definition file.

zhshize committed:
Update README.md doc for smooth writing

 - Fix typos
 - Remove some unnecessary sample code

zhshize committed:
Add phpdoc for remember me mechanism

zhshize committed:
Rename route/oidc-auth.php to route/oidc_auth.php

xdavidwu commented:
Mark resolved threads as resolved (thread status, not by commenting) and I will do another review round.

zhshize committed:
Rename the remember parameter to auto_refresh in the guard

zhshize committed:
Use the term 'OpenID Connect Provider'

zhshize committed:
Fix typo in comments

zhshize committed:
Remove unnecessary code

xdavidwu commented:
this needs a rebase

xdavidwu commented:
Closing due to inactivity and lack of consensus.

To whom it may concern, discuss with maintainers before implementing such a big patchset.
